libp2p/connmanager.nim

```diff
@@ -53,28 +58,42 @@ type
     muxer: Muxer
     handle: Future[void]
 
+  ConnHolder = object
+    conn: Connection
+    evtHandle: Future[void]
+
   ConnManager* = ref object of RootObj
     maxConns: int
-    # NOTE: don't change to PeerInfo here
-    # the reference semantics on the PeerInfo
-    # object itself make it susceptible to
-    # copies and mangling by unrelated code.
-    conns: Table[PeerID, HashSet[Connection]]
+    maxPeerConns: int
+    connSemaphore*: AsyncSemaphore
     muxed: Table[Connection, MuxerHolder]
+    conns: seq[ConnHolder]
     connEvents: Table[ConnEventKind, OrderedSet[ConnEventHandler]]
     peerEvents: Table[PeerEvent, OrderedSet[PeerEventHandler]]
 
 proc newTooManyConnections(): ref TooManyConnections {.inline.} =
   result = newException(TooManyConnections, "too many connections for peer")
 
 proc init*(C: type ConnManager,
-           maxConnsPerPeer: int = MaxConnectionsPerPeer): ConnManager =
-  C(maxConns: maxConnsPerPeer,
-    conns: initTable[PeerID, HashSet[Connection]](),
-    muxed: initTable[Connection, MuxerHolder]())
+           maxConns: int = MaxConnections,
+           maxPeerConns: int = MaxConnectionsPerPeer): ConnManager =
+  C(maxConns: maxConns,
+    maxPeerConns: maxPeerConns,
+    muxed: initTable[Connection, MuxerHolder](),
+    connSemaphore: AsyncSemaphore.init(maxConns))
 
 proc connCount*(c: ConnManager, peerId: PeerID): int =
-  c.conns.getOrDefault(peerId).len
+  c.conns
+  .filter((h) =>
+      (not isNil(h.conn.peerInfo) and
+      h.conn.peerInfo.peerId == peerId)
+  ).len
+
+proc peerCount*(c: ConnManager): int =
+  c.conns
+  .filter((c) => not isNil(c.conn.peerInfo))
```

Comment:
should unidentified peers ever be allowed inside the connection manager?

Reply:
yes, that allows us to properly track their counts and dispose of their resources if they close prematurelly.